enhancements/per-node-configuration.md

```diff
@@ -0,0 +1,850 @@
+  hostsession:
+    asn: 64512
+    hostasn: 64710
+---
+# L3Passthrough for internal nodes
+apiVersion: openpe.openperouter.github.io/v1alpha1
+kind: L3Passthrough
+metadata:
+  name: internal-passthrough
+  namespace: openperouter-system
+spec:
+  nodeSelector:
+    matchLabels:
+      security-zone: internal
+  hostsession:
+    asn: 64512
+    hostasn: 64720
+```
+
+#### Multiple Instances on Same Node
+
+Example showing multiple L3VNIs configured on the same set of nodes for different tenants:
+
+```yaml
+# Tenant A VNI on compute nodes
+apiVersion: openpe.openperouter.github.io/v1alpha1
+kind: L3VNI
+metadata:
+  name: tenant-a-vni
+  namespace: openperouter-system
+spec:
+  nodeSelector:
+    matchLabels:
+      node-role.kubernetes.io/worker: ""
+  vrf: tenant-a
+  vni: 5001
+  hostsession:
+    asn: 64512
+    hostasn: 64600
+---
+# Tenant B VNI on the same compute nodes
+apiVersion: openpe.openperouter.github.io/v1alpha1
+kind: L3VNI
+metadata:
+  name: tenant-b-vni
+  namespace: openperouter-system
+spec:
+  nodeSelector:
+    matchLabels:
+      node-role.kubernetes.io/worker: ""
+  vrf: tenant-b
+  vni: 5002
+  hostsession:
+    asn: 64512
+    hostasn: 64601
+---
+# Tenant C VNI on the same compute nodes
+apiVersion: openpe.openperouter.github.io/v1alpha1
+kind: L3VNI
+metadata:
+  name: tenant-c-vni
+  namespace: openperouter-system
+spec:
+  nodeSelector:
+    matchLabels:
+      node-role.kubernetes.io/worker: ""
+  vrf: tenant-c
+  vni: 5003
+  hostsession:
+    asn: 64512
+    hostasn: 64602
+```
+
+In this example, all compute nodes (with `node-role.kubernetes.io/worker` label) will have three L3VNIs configured, one for each tenant. This is allowed and expected behavior for L3VNI, L2VNI, and L3Passthrough resources.
+
+### Controller Implementation
+
+#### Node Matching Logic
+
+All controllers (Underlay, L3VNI, L2VNI, and L3Passthrough) will be enhanced with:
+
+1. **Node Label Watching**: Add a watch for Node resource label changes to trigger reconciliation
+2. **Selector Matching**: For each node, determine which resource instance applies by matching node labels against all resource `nodeSelector` fields
+3. **Configuration Application**: Generate and apply configuration only on nodes matched by the resource's selector
+4. **Dynamic Updates**: Reconcile when:
+   - Resources are created/updated/deleted
+   - Node labels change
+   - Nodes are added/removed from the cluster
+
+#### Conflict Resolution
+
+**Underlay-specific restriction:**
+
+Only one Underlay can match a given node at any time. The validation webhook will prevent conflicts at admission time:
+
+1. List all existing Underlay resources
+2. For each new/updated Underlay, check if any nodes would match multiple Underlays
+3. Reject the operation if overlapping selectors are detected
+
+**L3VNI, L2VNI, and L3Passthrough:**
+
+Multiple instances of L3VNI, L2VNI, and L3Passthrough can match the same node. This is intentional and allows for:
```

Comment:
this is true, but is not about conflict resolution. 
We need to ensure we don't have incompatible *vnis and passthrough on the same node (for each node)

Reply:
But this is already implemented right ? like right now if you for example have two l2vnis with the same vrf they fail, so is all about limiting this for example for l2vnis with the same node selectors.